### PR TITLE
[SPARK-24860][SQL] Support setting of partitionOverWriteMode in output options for writing DataFrame

### PR DESCRIPTION
**Pull request by koertkuipers**
## What changes were proposed in this pull request?

Besides spark setting spark.sql.sources.partitionOverwriteMode also allow setting partitionOverWriteMode per write 

## How was this patch tested?

Added unit test in InsertSuite

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 1]
```diff
@@ -91,8 +92,12 @@ case class InsertIntoHadoopFsRelationCommand(
...
     val pathExists = fs.exists(qualifiedOutputPath)
 
-    val enableDynamicOverwrite =
-      sparkSession.sessionState.conf.partitionOverwriteMode == PartitionOverwriteMode.DYNAMIC
+    val parameters = CaseInsensitiveMap(options)
+
+    val enableDynamicOverwrite = parameters.get("partitionOverwriteMode")
+      .map(mode => PartitionOverwriteMode.withName(mode.toUpperCase))
+      .getOrElse(sparkSession.sessionState.conf.partitionOverwriteMode) ==
+      PartitionOverwriteMode.DYNAMIC
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: nit: this is too long
```
val partitionOverwriteMode = parameters.get("partitionOverwriteMode")....
val enableDynamicOverwrite = partitionOverwriteMode == PartitionOverwriteMode.DYNAMIC
```

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 2]
```diff
@@ -486,6 +486,25 @@ class InsertSuite extends DataSourceTest with SharedSQLContext {
       }
     }
   }
+  test("SPARK-24860: dynamic partition overwrite specified per source without catalog table") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: nit: add a blank file before this new test case

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 3]
```diff
@@ -486,6 +486,25 @@ class InsertSuite extends DataSourceTest with SharedSQLContext {
...
+    withTempPath { path =>
+      Seq((1, 1, 1)).toDF("i", "part1", "part2")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: can we simplify the test? ideally we only need one partition column, and write some initial data to the table. then do an overwrite with partitionOverwriteMode=dynamic, and another overwrite with partitionOverwriteMode=static.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -1335,7 +1335,9 @@ object SQLConf {
         "overwriting. In dynamic mode, Spark doesn't delete partitions ahead, and only overwrite " +
         "those partitions that have data written into it at runtime. By default we use static " +
         "mode to keep the same behavior of Spark prior to 2.3. Note that this config doesn't " +
-        "affect Hive serde tables, as they are always overwritten with dynamic mode.")
+        "affect Hive serde tables, as they are always overwritten with dynamic mode. This can " +
+        "also be set as an output option for a data source using key partitionOverwriteMode, " +
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Also need to explain the precedence between the option and sqlconf.

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master

**Pull request closed without merging**